### PR TITLE
Add default retrieve limit

### PR DESCRIPTION
**Pull request by xige-16**
issue: #24435 

/kind bug
Signed-off-by: xige-16 <xi.ge@zilliz.com>

### CODE REVIEW

File: internal/core/src/segcore/SegmentInterface.cpp [hunk 1]
```diff
@@ -77,13 +77,35 @@ SegmentInternalInterface::Search(
 
 std::unique_ptr<proto::segcore::RetrieveResults>
 SegmentInternalInterface::Retrieve(const query::RetrievePlan* plan,
-                                   Timestamp timestamp) const {
+                                   Timestamp timestamp,
+                                   int64_t limit_rows,
```
Comment (czs007) [inline internal/core/src/segcore/SegmentInterface.cpp:81]: There is no need to specify a limit for the number of rows.

File: internal/querynodev2/segments/segment_loader.go [hunk 2]
```diff
@@ -586,6 +592,18 @@ func (loader *segmentLoader) loadFieldsIndex(ctx context.Context, segment *Local
...
 
 		segment.AddIndex(fieldID, fieldInfo)
+
+		// set average row data size of variable field
+		field, err := schemaHelper.GetFieldFromID(fieldID)
+		if err != nil {
+			return err
+		}
+		if typeutil.IsVariableDataType(field.GetDataType()) {
+			err = segment.UpdateFieldRawDataSize(numRows, fieldInfo.FieldBinlog)
```
Comment (longjiquan) [inline internal/querynodev2/segments/segment_loader.go:602]: OK, I got it. When the field was indexed, we use binlog size to set the average size.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/core/src/segcore/SegmentInterface.h [hunk 3]
```diff
@@ -69,7 +71,15 @@ class SegmentInterface {
     get_deleted_count() const = 0;
 
     virtual int64_t
-    get_real_count() const = 0;
+    get_real_count() = 0;
+
+    virtual int64_t
+    get_field_avg_size(FieldId field_id) = 0;
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/src/segcore/SegmentInterface.h:77]: Also, this function should be marked as const.

File: internal/core/src/segcore/SegmentInterface.h [hunk 4]
```diff
@@ -53,7 +53,9 @@ class SegmentInterface {
            Timestamp timestamp) const = 0;
 
     virtual std::unique_ptr<proto::segcore::RetrieveResults>
-    Retrieve(const query::RetrievePlan* Plan, Timestamp timestamp) const = 0;
+    Retrieve(const query::RetrievePlan* Plan,
+             Timestamp timestamp,
+             int64_t limit_size = DEFAULT_MAX_OUTPUT_SIZE) = 0;
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/segcore/SegmentInterface.h:58]: > Default parameter value is not suggested for virtual function.

Why？

File: internal/core/src/segcore/SegmentInterface.cpp [hunk 5]
```diff
@@ -178,4 +189,56 @@ SegmentInternalInterface::get_real_count() const {
...
 }
 
+int64_t
+SegmentInternalInterface::get_field_avg_size(FieldId field_id) {
+    if (SystemProperty::Instance().IsSystem(field_id)) {
+        if (field_id == TimestampFieldID || field_id == RowFieldID) {
+            return sizeof(int64_t);
+        }
+
+        throw std::runtime_error("unsupported system field id");
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/segcore/SegmentInterface.cpp:199]: > Add ut for this case.

see internal/core/unittest/test_retrieve.cpp：271

File: internal/core/src/segcore/segment_c.h [hunk 6]
```diff
@@ -103,6 +104,12 @@ CStatus
 UpdateSealedSegmentIndex(CSegmentInterface c_segment,
                          CLoadIndexInfo c_load_index_info);
 
+CStatus
+updateFieldRawDataSize(CSegmentInterface c_segment,
```
Comment (longjiquan) [inline internal/core/src/segcore/segment_c.h:108]: Capitalize the first letter to keep same coding style with other apis.

**Review by longjiquan — CHANGES_REQUESTED** [review]

**Pull request merged**